Header plugin

gsouf opened the pull request:
document https://github.com/php-http/plugins/pull/44

dbu commented:
needs to be updated once https://github.com/php-http/plugins/pull/44 has been adjusted

gsouf commented:
Yes I will wait for the final validation!

gsouf commented:
Please can you review it and tell me if it's clear enough? Maybe do we want to do

input => output example? if yes I dont know how to make it clear

dbu commented on plugins/header.rst:
```diff
@@ -0,0 +1,121 @@
+    // If a request handled by this client doesn't have a 'User-Agent' header,
+    // then $defaultUserAgent will be used
+    // If a request handled by this client has already a 'User-Agent' header,
+    // then the request wont be changed
```
imho this comment is too much. its already very well explained above

dbu commented on plugins/header.rst:
```diff
@@ -0,0 +1,121 @@
+
+
+    // Every request handled by this client will have the given 'User-Agent' and 'Accept' headers values
+    // If the request already has one of these header then it will be overridden with the given value
```
same, comment not needed

dbu commented on plugins/header.rst:
```diff
@@ -0,0 +1,121 @@
+    $userAgent = 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:40.0) Gecko/20100101 Firefox/40.1';
+
+    $headerRemovePlugin = new HeaderRemovePlugin([
+        'User-Agent' => $userAgent
```
this is wrong, no? the argument should be a simple array with value 'User-Agent'.

gsouf replied:
You are right, copy/paste mistake!

dbu commented on plugins/header.rst:
```diff
@@ -0,0 +1,121 @@
+    $myIp = '100.100.100.100';
+
+    $headerAppendPlugin = new HeaderAppendPlugin([
+        'X-Forwarded-For' => $myIp
```
the rfc define a Forwarded header - lets use that rather than a non-standard header (i know the x-header is widely used, but still)

gsouf replied:
I'm not sure about the Forwarded header. Should I do: ``['forwarded' => 'for=' . $myIp]`` ?

dbu replied:
i think that is correct

dbu commented on plugins/header.rst:
```diff
@@ -0,0 +1,121 @@
+Header Plugin
```
i think it makes sense to group those plugins in one doc page. but filename and title should be plural: header plugins

dbu commented:
good job, i like this doc, find it clear and understandable. you need to add this to the table of contents (look for other plugin filenames without the .rst extension and do the same)

gsouf commented:
Updated the doc, I also added a final example mixing up 2 header plugins. The goal is to make sure things are clear for everyone. Tell me if it's ok to keep it

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+Default headers values
+----------------------
+
+The plugin Http\Plugins\HeaderDefaultPlugin allows to set default values for given headers.
```
I don't think a FQCN is required here. Or is it?

gsouf replied:
Indeed it can feel weird for user, and the FQCN is specified in the code example

dbu replied:
i like the way joel is doing this in #76, with ``HeaderDefaultPlugin`` (local class in double backticks to make it shown as code. FQN in the code example is enough

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+----------------------
+
+The plugin Http\Plugins\HeaderDefaultPlugin allows to set default values for given headers.
+That means if the request hasn't already a given header it will be set,
```
does not already have?

gsouf replied:
Not sure but I think both are correct, though I'm not a pro with english!

dbu replied:
That means if a header is not set on a request, it will be added. However, if the header already is present on the request, the request is left unchanged.

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+
+The plugin Http\Plugins\HeaderDefaultPlugin allows to set default values for given headers.
+That means if the request hasn't already a given header it will be set,
+but if the request already has a value for the given header then the request wont be changed
```
won't

sagikazarmark replied:
Missing period

gsouf replied:
@sagikazarmark what do you mean by ``period``

sagikazarmark replied:
I mean a missing dot from the end of the sentence.

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+------------------------
+
+The Http\Plugins\HeaderSetPlugin allows to fix values of given header. That mean that any request passing through
+this plugin will have the same values for given headers
```
same? I think this is confusing. It means that each request for the given headers, but can also mean the same value for given headers

sagikazarmark replied:
Also, missing period

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+Mandatory headers values
+------------------------
+
+The Http\Plugins\HeaderSetPlugin allows to fix values of given header. That mean that any request passing through
```
I means that?

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+----------------
+
+The Http\Plugins\HeaderRemovePlugin allows to remove given headers from the request.
+That mean that any request passing through this plugin will be freed of the given headers.
```
It means that?

sagikazarmark replied:
Also the two sentences seem to be somewhat identical to me. I see that it matches the above two, but I think it would be better to somehow make them into one.

gsouf replied:
Yes I can i wanted to make things very clear, but the "remove" is already clear enough I think

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+-----------------------
+
+The Http\Plugins\HeaderAppendPlugin allows to set headers or to add values to existing headers.
+That means if the request has already the given headers then the value will be appended to the current value
```
already has

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+
+The Http\Plugins\HeaderAppendPlugin allows to set headers or to add values to existing headers.
+That means if the request has already the given headers then the value will be appended to the current value
+but if the request hasn't already the given header the header will be added to the request with the given value.
```
does not already have?

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+
+.. note::
+
+    This header is very specific and is mostly useful for headers like "forwarded"
```
Isn't it X-Forwarded?

gsouf replied:
Not sure about this, but in the previous revision I used ``X-Forwarded-For`` and @dbu asked me if I could use ``Forwarded``instead because it's the one from the RFC: https://tools.ietf.org/html/rfc7239#section-4

sagikazarmark replied:
Ok

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+Mixing operations
+-----------------
+
+Different header plugin can be mixed to achieve different operations
```
Different header plugins

sagikazarmark replied:
to achieve different behavior?

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+-----------------
+
+Different header plugin can be mixed to achieve different operations
+and you can use the same plugin for identical operations
```
Missing period

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,143 @@
+Different header plugin can be mixed to achieve different operations
+and you can use the same plugin for identical operations
+
+The following example will  the force the "User-Agent" and the "Accept" header values while removing "Cookie" header:
```
extra space

gsouf replied:
And extra "the" :D

sagikazarmark commented:
Good job @gsouf :+1: 

I added some semantical remarks, can you please take a look at them? Some are rather just suggestions, not actual corrections, because in some cases it is not necessarily incorrect, but feels a bit weird.

gsouf commented:
I added asked a few questions in the comments, can you check them please? Then I will fix the PR

sagikazarmark commented:
Questions answered :smile:

gsouf commented:
Just updated, sorry for the delay

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,142 @@
+
+The plugin ``HeaderDefaultPlugin`` allows to set default values for given headers.
+That means if a header is not set on a request, it will be added.
+However, if the header already is present on the request, the request is left unchanged.
```
I think "on the request" is not necessary

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,142 @@
+-----------------------
+
+The ``HeaderAppendPlugin`` allows to set headers or to add values to existing headers.
+That means if the request already has  the given headers then the value will be appended to the current value
```
extra space after has

sagikazarmark commented on plugins/headers.rst:
```diff
@@ -0,0 +1,142 @@
+
+The ``HeaderAppendPlugin`` allows to set headers or to add values to existing headers.
+That means if the request already has  the given headers then the value will be appended to the current value
+but if the request does not already have the given header the header will be added to the request with the given value.
```
s/the header/it

dbu replied:
and a missing comma. so maybe: "... does not already have the given header, it will be added with the given value."

sagikazarmark commented:
Added a few more notes. Apart from those, LGTM. @dbu anything from your side?

dbu commented on plugins/headers.rst:
```diff
@@ -0,0 +1,142 @@
+
+.. note::
+
+    This header is very specific and is mostly useful for headers like "forwarded"
```
"this header"? or rather "this plugin"?

dbu commented:
added a bit more nitpicking, but looks good to me overall.

gsouf commented:
done!

dbu commented:
thanks a lot!

gsouf commented:
You're welcome :)
